### PR TITLE
testbench server: start on an available port

### PR DESCRIPTION
**Pull request by mightyiam**

### CODE REVIEW

File: test/bench/server/index.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
-var startServer = require('./start-server')
+var server = require('./server')
+var getPort = require('./get-port')
+var listen = require('./listen')
 
-startServer()
+module.exports = getPort(server).then(listen)
```
Comment (mightyiam) [inline test/bench/server/index.js]: The testbench server now exports a promise that fulfills to the state of the running server.

**Comment by mightyiam** [comment]
With this merged in we can start writing a functional tests setup.

File: test/bench/server/index.js [hunk 2]
```diff
@@ -1,3 +1,16 @@
-var startServer = require('./start-server')
+var setup = require('./setup')
+var getPort = require('./get-port')
+var listen = require('./listen')
```
Comment (mightyiam) [inline test/bench/server/index.js]: Three steps...

File: test/bench/server/index.js [hunk 3]
```diff
@@ -1,3 +1,16 @@
...
+ * of the running testbench server.
+ *
+ * Properties are:
+ *  * `server` the express Application instance
+ *  * `port` the port on which it is listening
+ *
+ *  This module can be required or run standalone
+ *  (`node test/bench/server`).
+ */
+module.exports = setup.then(getPort).then(listen)
```
Comment (mightyiam) [inline test/bench/server/index.js]: Three steps.

**Comment by mightyiam** [comment]
@winhamwr do you have experience with Promises? I'm wondering how right I got this.

**Comment by winhamwr** [comment]
> do you have experience with Promises? I'm wondering how right I got this.

Only using them. I've never written them. It looks sane to me, though.

**Pull request merged**